### PR TITLE
Set each page flags directly instead of only setting the specific bit

### PR DESCRIPTION
**Pull request by ahrtr**
Each page flag should be exactly matching the expected flags instead of just specific bit equal.

cc @ptabor

### CODE REVIEW

**Comment by serathius** [comment]
+1, but leaving it to @ptabor to approve.

**Comment by cenkalti** [comment]
Flags field is a 16-bit integer. Today we only have page type flags that only occupies 4 bits which are mutually exclusive. However, we can use other 12 bits for other purposes in future. If we merge this, we'll lose that ability.

One quick idea:
- `bbolt check` command may set a flag when it detects a corruption in page, bbolt.Open may be opened with an option that skips corrupt pages based on the presence of corrupt-bit, allowing the user to copy existing data and recover from the situation.

**Comment by cenkalti** [comment]
Another quick idea:
- If we decide to change the page structure one day and decide to add extra fields, we can encode version into this field without breaking backwards compatibility.

**Comment by ahrtr** [comment]
Note this PR doesn't change the behavior. It just makes more explicit what it's currently doing.

Enhancing the field to contain more info is an easy & backward compatible task; but the reverse direction isn't. So I don't think it's a big problem, let alone we don't have any clear plan to change it for now.

**Comment by ahrtr** [comment]
cc @ptabor

**Comment by ptabor** [comment]
I'm supporting sentiment towards keeping this bits for future use in general, 
but In practice I don't seen and short-mid term solution to use this bits, so I'm fine with "freezing" this field for page-type.

Reasons for not expanding:
1. Unfortunately I already broke that compatibility in:  https://github.com/etcd-io/bbolt/commit/3cbd9c9a4449793de6506101056634efc11aaad8, 
it is v1.3.7 will not able to cooperate with any files that was written by a newer version that utilizes these bits. 

2. Would introducing a meaning to this bits require bump of: 
https://github.com/etcd-io/bbolt/blob/053fd189c85cc19cada191b10922ca944343194c/internal/common/types.go#L13
  1) If yes -> than it means that no old bbolt can run on these files as we validate the meta page version.
  2) If no (it is, we assume this field represents major changes to the page layout and introducing of new meaning of bits is OK) --> this means we need to be ready for silently dropping/messing this bits by old versions... that is very error prone.

I think that if we ever wanted to use this additional bits we would need: 
  1) Release new version (with storage format Version=3) that is not backward compatible with V2 (i.e. cannot be opened by old versions). 
  2) The new format would need to introduce proper Major.Minor versioning. 
  3) Within Minor version we could allow backward compatible bit's extensions assuming Major version preserves unknown bits.  But this allow us to entirely rethink the structure for extensibility. 

In scope of Version=2 I think we are locked to the current meaning of the field.

**Review by ptabor — APPROVED** [review]

**Comment by ahrtr** [comment]
Thanks @ptabor for the comment, which overall makes sense to me. 

> Release new version (with storage format Version=3) that is not backward compatible with V2

- I think new release library (e.g. 3) should be compatible with old version data file (e.g. 2), 
- Regard to running old release library (e.g. 2) on new version data file (e.g. 3), we can either fail it directly or ignore the new features with warning message. 

Since we don't have a detailed plan to extend the field (e.g reuse other bits) for now, so we can have more detailed discussion when we plan to extend it in future.

**Comment by cenkalti** [comment]
If the `Page.flags` fields is going to be locked to page type, can we rename the field as `Page.pageType` to make the intention clear?

**Comment by ahrtr** [comment]
> If the `Page.flags` fields is going to be locked to page type, can we rename the field as `Page.pageType` to make the intention clear?

Thanks for the suggestion, but I do not see a strong reason to rename it. 
- `flags` is just a flag, and we have a method `(*Page) Typ()` to map the flags to a human-readable page type, and methods `IsXXXPage()` to check each page type;
- There isn't a good name either even we want to rename it. `type` is a keyword in golang, and `pageType` is too wordy. It's already a field of struct `Page`, no need to add prefix `page` for the field again. 

Anyway, let's keep it as it's.
